src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -405,13 +481,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+            FreeDeconstructionVariables(variables);
+
+            return result;
+        }
+
+        /// <summary>
+        /// The caller is responsible for releasing the nested ArrayBuilders.
+        /// </summary>
+        private ArrayBuilder<DeconstructionVariable> BindDeconstructionDeclarationVariables(VariableDeclarationSyntax node, TypeSyntax varSyntax, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.Deconstruction != null);
+            var variables = node.Deconstruction.Variables;
+
+            if (varSyntax == null)
+            {
+                varSyntax = node.Type;
+            }
+
+            // There are four cases for VariableDeclaration:
+            // - type and declarators are set, but deconstruction is null. This could represent `int x`, which is a single variable.
+            // - type is null, declarators are set, but deconstruction is null. This could represent `x`, which is a single variable.
+            // - type is set to 'var', declarators are null, and deconstruction is set. This could represent `var (...)`
+            // - type and declarators are null, but deconstruction is set. Thi could represent `(int x, ...)`
+
+            var localsBuilder = ArrayBuilder<BoundLocalDeclaration>.GetInstance(variables.Count);
+            var variablesBuilder = ArrayBuilder<DeconstructionVariable>.GetInstance(variables.Count);
+            foreach (var variable in variables)
+            {
+                if (variable.Deconstruction == null)
+                {
+                    variablesBuilder.Add(new DeconstructionVariable(BindDeconstructionDeclarationVariable(variable, varSyntax, localsBuilder, diagnostics)));
+                }
+                else
+                {
+                    variablesBuilder.Add(new DeconstructionVariable(BindDeconstructionDeclarationVariables(variable, varSyntax, diagnostics)));
+                }
+            }
+
+            return variablesBuilder;
+        }
+
+        /// <summary>
+        /// Returns a BoundLocal when the type was explicit, otherwise returns a DeconstructionLocalPendingInference.
+        /// </summary>
+        private BoundExpression BindDeconstructionDeclarationVariable(VariableDeclarationSyntax node, TypeSyntax varSyntax, ArrayBuilder<BoundLocalDeclaration> locals, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.Deconstruction == null);
+            Debug.Assert(varSyntax != null || node.Type != null);
+            Debug.Assert(node.Variables.Count == 1);
+
+            var declarator = node.Variables[0];
+            var typeSyntax = varSyntax ?? node.Type;
+
+            bool isConst = false;
+
+            AliasSymbol alias;
+            bool isVar;
+            TypeSymbol declType = BindVariableType(node, diagnostics, typeSyntax, ref isConst, out isVar, out alias);
+            Debug.Assert(isVar || varSyntax == null); // if varSyntax is set, then isVar is true
```

Comment:
I'm not sure which fix caused the change, but I am now hitting that assert ;-)
I'll remove it.